### PR TITLE
Improve Rational instance

### PR DESCRIPTION
**Pull request by treeowl**

### CODE REVIEW

**Comment by treeowl** [comment]
Mostly unrelatedly: I'd really like to expose both `positiveRationals` and a version thereof that uses `Ratio Natural` instead of `Ratio Integer`. Any thoughts on where such should go? Also, I really think `Ratio Natural` deserves a `Universe` instance. I know that'll hamper type inference somewhat, but it feels like the right thing to do.

File: instances/base/Data/Universe/Instances/Base.hs [hunk 1]
```diff
@@ -59,24 +61,69 @@ instance Universe a => Universe (Last    a) where universe = map Last    univers
...
--- the same efficiency for generating the entire list (e.g. the times for
--- finding the sum of the first 100000 rationals are pretty much
--- indistinguishable).  positiveRationals is still the clear winner for
--- generating just the nth rational for some particular n -- some simple
--- experiments seem to indicate that doing this with positiveRationals2
--- scales linearly while with positiveRationals it scales sub-linearly,
--- as expected.
+-- Compiling this code with -O2 and doing some informal tests seems to show
+-- that positiveRationals isn't terribly much more efficient than
+-- positiveRationals for generating the entire list (e.g. the times for finding
```
Comment (alexfmpe) [inline instances/base/Data/Universe/Instances/Base.hs]: Missed a '2'?

Comment (treeowl) [reply instances/base/Data/Universe/Instances/Base.hs]: Yeah, I'll fix that comment.

**Comment by phadej** [comment]
Please rebase on top of current `master`, the travis is green there.

**Comment by phadej** [comment]
How about

```haskell
class UniverseRatio a where
   positiveRationalStream :: Stream (Ratio a)
   rationalStream :: Stream (Ratio a)

instance UniverseRatio Integer where ...
instance UniverseRatio Natural where ...

instance UniverseRatio a => Universe (Ratio a) where universe = ... rationalStream

**Comment by treeowl** [comment]
Hrmmmm.... Can we do better? We should probably be able to get more sharing by taking advantage of the fact that each row of the Calkin-Wilf tree consists of a finite sequence and its reversed reciprocal. That could even win some sharing on the negative side. I don't know if that will actually help in practice.

**Comment by treeowl** [comment]
@phadej, that was the sort of thing I was talking about. If we wanted, we could probably also add instances for `Word64` and `Int64` using the same construction; they shouldn't overflow until somewhere around the 2^64th element (the `fusc` function maxes out around the identity function), which seems good enough for most real calculations. To get the exact full sequences for bounded integral types, we could probably be a bit careful and check for overflow.

**Comment by phadej** [comment]
`Int64` there should be a proper justification (both correctness, and practical value), or let's leave it out.

**Comment by treeowl** [comment]
How tied are we to the current order? I strongly suspect we can get some improvements by changing it a bit, interleaving among the intervals (-∞, -1), (-1, 0), (0, 1), and (1,  ∞). Basically, we can generate coprime pairs `(m,n)` with `m < n` (the left side of the Calkin-Wilf tree) and from each produce `m / n`, `n / m`, `-m / n`, and  `-n / m`.

**Comment by treeowl** [comment]
Er, sorry, I guess I was wrong about `m < n`; that's not guaranteed, but it's also not important.

**Comment by treeowl** [comment]
Yeah, changing the order seems to make quite a big difference. I think we should do it.

**Comment by treeowl** [comment]
@dmwit, are you okay with my proposed reordering? My informal tests suggests it helps a lot.

**Comment by phadej** [comment]
Merged as part of #41

**Pull request closed without merging**